tests/test_training_simple.py

```diff
@@ -14,3 +42,265 @@ def test_train_simple():
+
+
+    # Now do training...
+    if args.distributed:
+        dist.barrier()
+
+    success, global_step = train_one_epoch(
+        main_vars['model'],
+        main_vars['data'],
+        main_vars['loss'],
+        epoch=main_vars['start_epoch'],
+        step=main_vars['global_step'],
+        optimizer=main_vars['optimizer'],
+        scaler=main_vars['scaler'],
+        scheduler=main_vars['scheduler'],
+        total_steps=main_vars['total_steps'],
+        args=main_vars['args'],
+        tb_writer=main_vars['writer'],
+    )
+    if args.distributed:
+        dist.barrier()
+
+    lrs = [_['lr'] for _ in main_vars['optimizer'].param_groups]
+
+    target = 0.5 * (1 + np.cos(np.pi * (num_batches-1 - args.warmup) / (num_batches - args.warmup))) * args.lr
+    assert all(abs(lr-target) < 1e-10 for lr in lrs)
+
+
+
+def test_lr_multi_epoch_cyclic():
+    """ Tests that LR gets adjust correctly for a single epoch (that overruns the warmup)
+    """
+    seq_len = 16
+    batch_size = 2
+    num_epochs = 5
+    total_batches = 1000
+    num_batches = total_batches // num_epochs
+    print("NUM BATCHES", num_batches)
+    args = ['--train-num-samples', str(num_batches * seq_len * batch_size),
+            '--batch-size', str(batch_size),
+            '--dataset-type', 'synthetic',
+            '--model', 'open_lm_test_tiny',
+            '--epochs', str(num_epochs),
+            '--finegrain-debug',
+            '--lr', '1e0',
+            '--warmup', str(100)] # short warmup
+    main_vars = main(args)
+
+    args = main_vars['args']
+    assert args.warmup == 100 # Default warmup steps
+    assert num_batches > args.warmup
+
+
+    compute_lr = lambda s: 0.5 * args.lr * (1 + np.cos(np.pi * (s - 100) / (total_batches - 100)))
+    # Now do training...
+    global_step = main_vars['global_step']
+    all_lrs = []
+    for epoch in range(num_epochs):
+        if args.distributed:
+            dist.barrier()
+
+        success, global_step = train_one_epoch(
+            main_vars['model'],
+            main_vars['data'],
+            main_vars['loss'],
+            epoch=main_vars['start_epoch'],
+            step=global_step,
+            optimizer=main_vars['optimizer'],
+            scaler=main_vars['scaler'],
+            scheduler=main_vars['scheduler'],
+            total_steps=main_vars['total_steps'],
+            args=main_vars['args'],
+            tb_writer=main_vars['writer'],
+        )
+        if args.distributed:
+            dist.barrier()
+
+        lrs = [_['lr'] for _ in main_vars['optimizer'].param_groups]
+
+        all_lrs.append(lrs)
+        # return args, all_lrs
+        if global_step >= 100:
+            assert all(abs(lr - compute_lr(global_step - 1)) < 1e-8 for lr in lrs)
+
+
+
+# =========================================================
+# =                main.py tests for LR                   =
+# =========================================================
+
+
+def test_lr_scheduling_from_main():
```

Comment:
Can we just use this style of tests for all tests (read from TB without finegrain_debug)? It'll simplify the params and main and we don't need finegrain_debug.